### PR TITLE
Ribbons

### LINKED ISSUE #9: Encode pitch variation (windowed standard deviation) of as ribbon width

### PR DESCRIPTION
**Pull request by curran**
Will close #9

### CODE REVIEW

**Comment by curran** [comment]
Current status:

image

**Comment by seemantk** [comment]
Starting to emerge. Nice progress, @curran!

**Comment by curran** [comment]
Little by little! Got colors working, but something is still horribly wrong.

image

**Comment by curran** [comment]
image

Making progress but still debugging how some NaNs are ending up in the path. Seems to be a problem only for the blue voice.

This is using the (min, max) for each window step, as a simple proxy for the standard deviation calculation.

**Comment by curran** [comment]
Aha! I realized where the NaNs were coming from. As the Y scale is a Band scale not a Linear scale, it returns NaN when given input values that were not explicitly set on its domain.

**Comment by curran** [comment]
Woohoo solved that tricky issue with the NaNs! Here's what we've got now:

image

**Comment by curran** [comment]
Smooth curves 
image

**Comment by curran** [comment]
Finally working! Here's what windowed standard deviation looks like.

image

**Comment by curran** [comment]
Cleaned up the code significantly, and addressed some edge cases where the notes were not considered as "inside the window" unless they were completely inside (rather than having just a portion inside). Now it uses a more inclusive definition, so a note with a long duration will be considered "inside" the window even if only a portion of it (beginning, middle, or end) is inside the window.

Here's what it looks like now:

image

@seemantk This is ready for review. I'd consider the core ribbon feature "done", however there are still some rough edges here, including:
- [ ] UI is needed for toggling the ribbons (currently they are always on).
- [ ] Transitions are not smooth (between split and combined).
- [ ] In separated mode when viewing a single voice, there are degenerate ribbons displayed for the other voices.
- [ ] In combined mode when viewing a single voice, the ribbons are all the same opacity, but the ribbons for the voices other than the selected voice should be faded out (as the notes are).

If you want to merge, we can track these remaining tasks as separate issues and tackle them one by one.

**Comment by curran** [comment]
Here's what the UI looks like now, after the addition of the check box for toggling the ribbons:

image
Checked

image
Unchecked

**Comment by curran** [comment]
We implemented the attack density computation using the algorithm discussed in #10. Here is [the implementation](https://github.com/sul-cidr/josquin-ribbon/pull/71/commits/99cdafc5161e35aab97fe326914f35120a6ec7c8#diff-21cc978204e78587b9d3af2109573897R124) for review.

Here's what the attack density ribbon looks like:

image

**Comment by curran** [comment]
Added radio buttons for switching the mode of the ribbon width:

image

When the "Show Ribbons" checkbox is unchecked, the radio buttons disappear:

image

@seemantk This is ready for review and merge.

File: js/notesbook.js [hunk 1]
```diff
@@ -84,6 +86,7 @@ function NotesBook() {
       barlines = g
         .append("g")
           .attr("class", "barlines")
+          .attr("clip-path", "url(#noteclip)")
```
Comment (seemantk) [inline js/notesbook.js:89]: <3 SVG!

File: js/ribbon.js [hunk 2]
```diff
@@ -24,35 +30,42 @@ function Ribbon() {
...
-            //    x: x
-            //  , y0: d3.max(notesInWindow, function (d){ return d.pitch; })
-            //  , y1: d3.min(notesInWindow, function (d){ return d.pitch; })
-            //};
 
             // This is a simple temporary proxy for the avg +- standard deviation.
-            var maxPitch = d3.max(notesInWindow, function (d){ return d.pitch; })
-            var minPitch = d3.min(notesInWindow, function (d){ return d.pitch; })
+            var extent = d3.extent(notesInWindow, function (d){ return d.pitch; })
+
```
Comment (seemantk) [inline js/ribbon.js]: nice use of d3.extent outside of usual scale things

File: js/ribbon.js [hunk 3]
```diff
@@ -23,42 +23,55 @@ function Ribbon() {
...
+            // at the notes in the window of width `interval`, centered on `x`.
+            var notesInWindow = data.value
+              .filter(function(d){
+                return (
+                  d.time > (x - interval / 2)
+                  &&
+                  (d.time + d.duration) < (x + interval / 2)
+                );
+              })
+              .map(function (d){ return d.pitch; })
```
Comment (seemantk) [inline js/ribbon.js]: Fantastic idea to move it here instead, saving CPU cycles.

File: js/ribbon.js [hunk 4]
```diff
@@ -23,42 +23,55 @@ function Ribbon() {
...
+                deviation = 0;
+                break;
+
+              // If there's only a single note in our window,
+              // then use its pitch as the mean, and a deviation of 0.
+              case 1:
+                mean = notesInWindow[0];
+                deviation = 0;
+                break;
+
```
Comment (seemantk) [inline js/ribbon.js]: Nice job sorting out the exceptions here

File: js/ribbon.js [hunk 5]
```diff
@@ -12,11 +12,17 @@ function Ribbon() {
...
           .curve(d3.curveBasis)
+      , ribbonData
+      , ribbonDataStale = true
     ;
 
     function my(){
       if(data && domain && scale){
-        var ribbonData = computeRibbon();
+
+        if(ribbonDataStale){
```
Comment (seemantk) [inline js/ribbon.js]: nice

File: js/ribbon.js [hunk 6]
```diff
@@ -46,24 +59,29 @@ function Ribbon() {
...
+              // if any part of it falls inside the window.
               return (noteStart < windowEnd) && (noteEnd > windowStart);
             })
             .map(function (d){ return d.pitch; })
           ;
 
+          // At each iteration of this function,
+          // we'll compute the mean and standard deviation.
+          var mean, deviation;
+
```
Comment (seemantk) [inline js/ribbon.js:81]: I really appreciate reading your explanation comments here and in other places you've done them.

File: js/ribbon.js [hunk 7]
```diff
@@ -38,6 +43,14 @@ function Ribbon() {
     }
 
     function computeRibbon(){
+      if(mode === STANDARD_DEVIATION){
+        return computeWindowedStandardDeviation();
+      } else if(mode === ATTACK_DENSITY){
+        return computeAttackDensity();
+      }
+    }
+
```
Comment (seemantk) [inline js/ribbon.js:49]: what do you think of putting this logic into a lookup table where the modes are defined? eg: `computeRibbon = { STANDARD_DEV: computeWinStdDev(), ATTACKS: computeAttackDensity() }`? then you can just run `computeRibbon[mode]`

File: js/ribbonsUI.js [hunk 8]
```diff
@@ -54,20 +58,30 @@ function RibbonsUI(){
...
         var radioLabel = radioContainer.selectAll("label")
-                .data(modes)
-              .enter().append("div").append("label")
-          , inputs = radioLabel
+            // Hide the radio buttons if the "show ribbons"
+            // box is not checked.
+            .data(showRibbons ? modes : []);
+
+        radioLabel.exit().remove();
+
```
Comment (seemantk) [inline js/ribbonsUI.js:68]: Nice touch to make this dynamic.

**Review by seemantk — APPROVED** [review]
Love what you've done with this branch, @curran.  The ribbon came together rather beautifully.

**Pull request merged**